AsyncNats/NatsServerPool.cs

```diff
@@ -0,0 +1,180 @@
+﻿namespace EightyDecibel.AsyncNats
+{
+    using System;
+    using System.Buffers;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+
+    internal class NatsServerPool:INatsServerPool
+    {
+        public List<DnsEndPoint> Servers => new List<DnsEndPoint>(_servers.Concat(_discoveredServers));
+
+        List<DnsEndPoint> _servers = new List<DnsEndPoint>();
+        List<DnsEndPoint> _discoveredServers = new List<DnsEndPoint>();
+        Random _random = new Random();
+        DnsEndPoint? _lastSelectedDnsEndPoint = null;
+        Queue<IPEndPoint> _retryIPEndPointQueue = new Queue<IPEndPoint>();
+        INatsOptions _options;
+        ILogger<NatsServerPool>? _logger;
+
+        object _sync = new object();
+
+        public NatsServerPool(INatsOptions options)
+        {
+
+            _logger=options.LoggerFactory?.CreateLogger<NatsServerPool>();
+
+            var servers = options.Servers;
+
+            if(servers == null) throw new ArgumentNullException(nameof(servers));
+            if(servers.Count()==0) throw new ArgumentException("server list cannot be empty",nameof(servers));
+
+            _options = options;
+
+            foreach (var server in servers)
+                AddServer(_servers,server);
+        }
+
+        public async ValueTask<IPEndPoint> SelectServer(bool isRetry=false)
+        {
+            DnsEndPoint SelectDnsEndpoint(bool isRetry)
+            {
+                DnsEndPoint selectedServer;
+                lock (_sync)
+                {
+                    var combinedServers = _servers;
+                    if (_options.ServersOptions.HasFlag(NatsServerPoolFlags.AllowDiscovery))
```

Comment:
I can offer you some real world example. Immagine you have a nats cluster sitting behind a load balancer. Each nats instance advertise their subnet ip ( 10.0.0.1 eg ) to the other members of the cluster. Now to allow access from outside, there's a dns hostname that resolves to the frontend ip of the load balancer. In such situation, if you are accessing the cluster from outside, you want to reject discovered servers since they'll contain an ip address that's private to a network you are not in. Fair enough to say such user should never enable the flag in this situation, but I believe it doesn't hurt to allow the misconfigured client to still function instead of throwing it at an unusable state constantly trying to connect to non-reachable addresses. I usually prefer the more forgiving behavior.